### PR TITLE
Add GitHub Actions CI

### PR DESCRIPTION
**Pull request by Margen67**
Based on #1717.
Relevant issue: #1848

Example run: https://github.com/Margen67/Cxbx-Reloaded/actions/runs/65162110
Example release: https://github.com/Margen67/Cxbx-Reloaded/releases/tag/CI-31d4ea3

### CODE REVIEW

File: .github/workflows/CI.yml [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+          cmake .. -A Win32
+      - name: Build
+        working-directory: build
+        run: cmake --build . --config ${{ matrix.configuration }} -j $env:NUMBER_OF_PROCESSORS
+      - name: Prepare artifacts
+        if: matrix.configuration == 'Release'
+        run: |
+          robocopy . artifacts COPYING README.md /r:0 /w:0
+          robocopy build\bin\${{ matrix.configuration }} artifacts Cxbx.exe glew32.dll subhook.dll SDL2.dll cxbxr-debugger.exe capstone.dll cs_x86.dll /r:0 /w:0
+          If ($LastExitCode -le 7) { $LastExitCode = 0 }
```
Comment (GXTX) [inline .github/workflows/CI.yml]: Why is this needed?

Comment (Margen67) [reply .github/workflows/CI.yml]: robocopy gives an error code even on success so it has to be suppressed or else it will be considered an error.

File: .github/workflows/CI.yml [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+      - uses: actions/upload-artifact@v1
+        if: matrix.configuration == 'Release'
+        with:
+          name: CxbxReloaded-${{ matrix.configuration }}-${{ matrix.vsver }}
+          path: artifacts
+
+  release:
+    if: github.event.action != 'pull_request' && github.ref == 'refs/heads/master' && github.repository == 'Cxbx-Reloaded/Cxbx-Reloaded'
```
Comment (GXTX) [inline .github/workflows/CI.yml]:
```
&& github.repository == 'Cxbx-Reloaded/Cxbx-Reloaded'
```

I don't see a need to keep the release from running *only* on the base repo, if users have forks and want the same functionality they'd have to modify the script.

**Comment by GXTX** [comment]
Overall looks good and seems to work as expected!

I brought up while we were chatting before I am sad that you're having to hard code the values for the release stage but i've yet to come up with a solution and isn't *that bad*.

The only other thing is the testing commits you've used but that can be dealt with by a squash on merge.

File: .github/workflows/CI.yml (hunk 2 above)
Comment (Margen67) [reply .github/workflows/CI.yml]: Having releases on all forks would be annoying for people that just want to make PRs.
Builds can still be downloaded from artifacts so releases aren't necessary in that case.

Comment (GXTX) [reply .github/workflows/CI.yml]: It would still only run on merges to `master`, so if people are using feature branches correctly it wouldn't run just for a PR.

**Comment by ergo720** [comment]
It would be better to squash all those commits into one.

File: .github/workflows/CI.yml (hunk 2 above)
Comment (Margen67) [reply .github/workflows/CI.yml]: That check is redundant. Removing it wouldn't be enough.
The `github.ref` check would have to be changed to something like `*/master`.

Also, I'm not sure if forking alone is enough to trigger GitHub Actions on the new repo, so that might be another reason not to have releases on forks aside from creating unneeded tags.

**Comment by Margen67** [comment]
AppVeyor needs to be paused, or #1846 needs to be merged to stop builds from triggering.

**Comment by RadWolfie** [comment]
In my opinion, CI builds and publish release should be in two separate files. I have a theory which may be a possible solution, I'm not sure if it will break at any time in the future. Have publish release yaml file in separate branch with only **one** commit. With this method, it should be easier and simple just to "delete" `release-ci` branch in forked repos.

---
I don't agree with mark CI builds as latest release. It should be mark as pre-release tag. Whilst `v#.#` tags will mark as release tag.

File: README.md [hunk 3]
```diff
@@ -1,5 +1,6 @@
 # Cxbx-Reloaded - Original Xbox Emulator
 [![License: GPL v2](https://img.shields.io/badge/License-GPL%20v2-blue.svg)](https://img.shields.io/badge/License-GPL%20v2-blue.svg)
+[![GitHub Actions](https://github.com/Cxbx-Reloaded/Cxbx-Reloaded/workflows/CI/badge.svg?branch=master)
```
[the remark this answers is not in the document]
Comment (RadWolfie) [reply README.md]: @GXTX I think it would be better to let @LukeUsher remove the badges whenever he feels ready to deactivate third-party CIs.

File: .github/workflows/CI.yml [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+    strategy:
+      fail-fast: false
+      matrix:
+        configuration: [Release, Debug]
+        vsver: [VS2019, VS2017]
+        include:
+          - vsver: VS2019
+            os: windows-latest
+          - vsver: VS2017
+            os: windows-2016
```
Comment (RadWolfie) [inline .github/workflows/CI.yml]: `windows-2016` is quite deprecated. GitHub dev team mentioned they will remove it in January 2020, then decided to postpone. I believe it is worth add a note if any time GitHub dev team finally remove `windows-2016` virtual environment, we should have a note ready to prepare what to do next.

Comment (GXTX) [reply .github/workflows/CI.yml]: Note: https://github.com/actions/virtual-environments/issues/68#issuecomment-602652021

File: README.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (GXTX) [reply README.md]: Okay I'm fine with that but the first part still needs to change.

**Comment by Margen67** [comment]
> 
> 
> In my opinion, CI builds and publish release should be in two separate files. I have a theory which may be a possible solution, I'm not sure if it will break at any time in the future. Have publish release yaml file in separate branch with only **one** commit. With this method, it should be easier and simple just to "delete" `release-ci` branch in forked repos.
> 
> I don't agree with mark CI builds as latest release. It should be mark as pre-release tag. Whilst `v#.#` tags will mark as release tag.

I still don't see why anyone would want releases when you can list runs and download artifacts.
The point of releases for master is to make a link to the latest build like so: `https://github.com/Cxbx-Reloaded/Cxbx-Reloaded/releases/latest/CxbxReloaded-Release-VS2019.zip`
Having releases for forks by default is stupid. People will have to remove it, and possibly delete the release+tag if it ran. There's a reason most things on GitHub don't do this.

Also, marking them as prerelease breaks the `latest` link.

**Comment by RadWolfie** [comment]
> I still don't see why anyone would want releases when you can list runs and download artifacts.
> The point of releases for master is to make a link to the latest build like so: `https://github.com/Cxbx-Reloaded/Cxbx-Reloaded/releases/latest/CxbxReloaded-Release-VS2019.zip`
> Also, marking them as prerelease breaks the `latest` link.

I rather send people to the release page. Plus there is a way to query releases' API for pre-release which can be done with javascript. GitHub design the latest release as a stable build usage. Except nothing is stable in our situation. [I had brought it up](https://github.com/Cxbx-Reloaded/Cxbx-Reloaded/issues/1848#issuecomment-601383511) and yet haven't been discussed in #1848 ticket.

> Having releases for forks by default is stupid. People will have to remove it, and possibly delete the release+tag if it ran. There's a reason most things on GitHub don't do this.

Right, it is better to have a separate yaml file just for (pre-)release upload which can only target to this repo. Forked repo users can update it on their side if they want to make a variant of Cxbx-Reloaded. Even continue with the project if we decide to stop maintenance Cxbx-Reloaded in the future.

**Comment by Margen67** [comment]
> 
> 
> > I still don't see why anyone would want releases when you can list runs and download artifacts.
> > The point of releases for master is to make a link to the latest build like so: `https://github.com/Cxbx-Reloaded/Cxbx-Reloaded/releases/latest/CxbxReloaded-Release-VS2019.zip`
> > Also, marking them as prerelease breaks the `latest` link.
> 
> I rather send people to the release page. Plus there is a way to query releases' API for pre-release which can be done with javascript. GitHub design the latest release as a stable build usage. Except nothing is stable in our situation. [I had brought it up](https://github.com/Cxbx-Reloaded/Cxbx-Reloaded/issues/1848#issuecomment-601383511) and yet haven't been discussed in #1848 ticket.
> 
> > Having releases for forks by default is stupid. People will have to remove it, and possibly delete the release+tag if it ran. There's a reason most things on GitHub don't do this.
> 
> Right, it is better to have a separate yaml file just for (pre-)release upload which can only target to this repo. Forked repo users can update it on their side if they want to make a variant of Cxbx-Reloaded. Even continue with the project if we decide to stop maintenance Cxbx-Reloaded in the future.

Wouldn't that require a website?

Releases aren't created for forks already, a separate yaml isn't needed for that.

File: .github/workflows/CI.yml (hunk 4 above)
Comment (Margen67) [reply .github/workflows/CI.yml]: It's only for WINE compatibility so hopefully WINE supports VS2019 runtimes before it gets removed.

Comment (GXTX) [reply .github/workflows/CI.yml]: Them closing the issue makes me think they're reevaluating it, I would think we still have some time before it actually gets removed.

**Comment by PatrickvL** [comment]
What's the status on this PR?

**Comment by LukeUsher** [comment]
Regarding artifacts: remember that GitHub actions does not currently allow non-registered users to download them, but releases are available to everyone.

**Comment by Margen67** [comment]
> 
> 
> What's the status on this PR?

It's ready to be merged if the commits don't matter.

File: .github/workflows/CI.yml [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+            cxbxr-debugger.exe capstone.dll cs_x86.dll /r:0 /w:0
+          If ($LastExitCode -le 7) { $LastExitCode = 0 }
+      - uses: actions/upload-artifact@v1
+        if: matrix.configuration == 'Release'
+        with:
+          name: CxbxReloaded-${{ matrix.configuration }}-${{ matrix.vsver }}
+          path: artifacts
+
+  release:
+    if: github.event.action != 'pull_request' && github.ref == 'refs/heads/develop' # TODO: Switch to master
```
Comment (RadWolfie) [inline .github/workflows/CI.yml:68]: Convert this line into multi-line and re-add `github.repository == 'Cxbx-Reloaded/Cxbx-Reloaded'` check.

File: .github/workflows/CI.yml [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+          7z a -mx1 $env:artifact-2.zip .\$env:artifact-2\*
+          echo "::set-env name=short_commit_sha::$(git rev-parse --short HEAD)"
+      - name: Create Release
+        id: create_release
+        uses: actions/create-release@v1
+        env:
+          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
+        with:
+          tag_name: CI-${{ env.short_commit_sha }}
+          release_name: CI-${{ env.short_commit_sha }}
```
Comment (RadWolfie) [inline .github/workflows/CI.yml:96]: Add `prerelease: true` after release_name line. All CI builds are not stable release nor is recommended. All published releases (including pre-releases) will always be at the top. Plus latest release tag does not stay on the top.

File: .github/workflows/CI.yml [hunk 7]
```diff
@@ -0,0 +1,114 @@
+name: CI
```
Comment (RadWolfie) [inline .github/workflows/CI.yml:1]: The badge itself will only show "CI | \<status\>", see live example: GitHub Actions

The name need to be clear from which source CI badge is coming from, disregard the GitHub's logo before "CI".

File: README.md [hunk 8]
```diff
@@ -1,5 +1,6 @@
 # Cxbx-Reloaded - Original Xbox Emulator
 [![License: GPL v2](https://img.shields.io/badge/License-GPL%20v2-blue.svg)](https://img.shields.io/badge/License-GPL%20v2-blue.svg)
+[![GitHub Actions](https://github.com/Cxbx-Reloaded/Cxbx-Reloaded/workflows/CI/badge.svg?branch=develop)
```
Comment (RadWolfie) [inline README.md:3]: The markdown need a fixup to include missing end bracket and link in the next parenthesis.

```suggestion
[![GitHub Actions](https://github.com/Cxbx-Reloaded/Cxbx-Reloaded/workflows/CI/badge.svg?branch=develop)](https://github.com/Cxbx-Reloaded/Cxbx-Reloaded/releases)
```

File: .github/workflows/CI.yml (hunk 7 above)
Comment (Margen67) [reply .github/workflows/CI.yml:1]: That's not possible.

File: .github/workflows/CI.yml (hunk 5 above)
Comment (Margen67) [reply .github/workflows/CI.yml:68]: I'm not sure if making it multi-line is possible.

`github.ref == 'refs/heads/develop'` already does acts as a repo check since forks won't have the same ref.

File: .github/workflows/CI.yml (hunk 7 above)
Comment (RadWolfie) [reply .github/workflows/CI.yml:1]: I didn't say the logo need to be remove, I'm aware it's a part of their badge system. I'm saying add "GitHub" text before CI aka
```suggestion
name: GitHub CI
```

File: .github/workflows/CI.yml (hunk 5 above)
Comment (RadWolfie) [reply .github/workflows/CI.yml:68]: Base on personal test, both base and forks has the same ref. Here's a working example:

- base
image
--
- fork
image

In conclusion, `github.ref == 'refs/heads/develop'` will become true for forks.

Comment (RadWolfie) [reply .github/workflows/CI.yml:68]: I have a working alternative solution for multi-line. Set the environments conditional, then use those environment names in if statement. Example:
```
if: env.isDevelopBranch && env.setRepo
env:
  isDevelopBranch: ${{github.ref}} == 'refs/heads/develop'
  setRepo: ${{github.repository}} == 'Cxbx-Reloaded/Cxbx-Reloaded'
```

File: .github/workflows/CI.yml (hunk 6 above)
Comment (GXTX) [reply .github/workflows/CI.yml:96]: > nor is recommended

I disagree, users essentially *only* use the "up-to-date" builds from PRs and when they come asking for help when using the "stable" release they're told by you and others to update to such things.

File: .github/workflows/CI.yml (hunk 7 above)
Comment (GXTX) [reply .github/workflows/CI.yml]: That gets messy if we add more workflows as each will have their own name, should this maybe be verbose? `Verify Builds` but then the badge says such.

File: .github/workflows/CI.yml (hunk 6 above)
Comment (RadWolfie) [reply .github/workflows/CI.yml:96]: That's because our release page doesn't have the development build pre-releases. There are people that goes on GitHub's release tab and only find "stable" release. They were not aware of development builds which is further down in readme file. Hence the question of "where's the latest build?" usually come from.

With GitHub Actions release to public, this kind of problem will be resolve.

File: .github/workflows/CI.yml [hunk 9]
```diff
@@ -64,28 +64,32 @@ jobs:
...
           path: artifacts
 
-  release: # TODO: Remove develop once rebased
-    if: github.event.action != 'pull_request' && github.ref == 'refs/heads/develop' || 'refs/heads/master' && github.repository == 'Cxbx-Reloaded/Cxbx-Reloaded'
+  release:
+    if: | # TODO: Remove develop once rebased
+      github.event.action != 'pull_request' &&
+      github.ref == 'refs/heads/develop'
+                 || 'refs/heads/master'     &&
+      github.repository == 'Cxbx-Reloaded/Cxbx-Reloaded'
```
Comment (RadWolfie) [inline .github/workflows/CI.yml:72]: Your multi-line commit didn't work as intended, I had verified this within my private Action-Test repo. Here's what you should change.
```suggestion
    if: | # TODO: Remove develop once rebased
      github.event.action != 'pull_request' &&
      (github.ref == 'refs/heads/develop' ||
       github.ref == 'refs/heads/master')     &&
      github.repository == 'Cxbx-Reloaded/Cxbx-Reloaded'
```

Comment (RadWolfie) [reply .github/workflows/CI.yml:72]: Otherwise, without above suggestion change. A new branch + add above code in a new commit will trigger job operation. If I add the change above, it will skip the job operation. Even skip it for a pull request to develop branch. Once merged, it will run the job. Again, make a pull request to master will not trigger it. Finally, merge pull request into master will trigger the job.
